crates/net/downloaders/src/headers/concurrent.rs

```diff
@@ -0,0 +1,599 @@
+            }
+
+            // update tracked block info
+            self.next_chain_tip_hash = parent.hash();
+            self.next_chain_tip_block_number = parent.number;
+            self.queued_validated_headers.push(parent);
+        }
+
+        Ok(())
+    }
+
+    /// Invoked when we received a response
+    #[allow(clippy::result_large_err)]
+    fn on_headers_outcome(
+        &mut self,
+        response: HeadersRequestOutcome,
+    ) -> Result<(), HeadersResponseError> {
+        let requested_block_number = response.block_number();
+        let HeadersRequestOutcome { request, retries, outcome } = response;
+
+        match outcome {
+            Ok(res) => {
+                let (peer_id, headers) = res.split();
+                let mut headers = headers.0;
+
+                // sort headers from highest to lowest block number
+                headers.sort_unstable_by_key(|h| Reverse(h.number));
+
+                if headers.is_empty() {
+                    return Err(HeadersResponseError {
+                        request,
+                        retries,
+                        peer_id: Some(peer_id),
+                        error: DownloadError::EmptyResponse,
+                    })
+                }
+
+                // validate the response
+                let highest = &headers[0];
+
+                if highest.number != requested_block_number {
+                    // TODO return error, invalid response
+                }
+
+                if (headers.len() as u64) != request.limit {
+                    // TODO decide how to treat this
+                }
+
+                // check if the response is the next expected
+                if highest.number == self.next_chain_tip_block_number {
+                    // is next response, validate it
+                    self.process_next_headers(request, headers, peer_id, retries)?;
+
+                    // try to validate all buffered responses blocked by this successful response
+                    self.try_validate_buffered()
+                        .map(Err::<(), HeadersResponseError>)
+                        .transpose()?;
+                } else {
+                    // can't validate yet
+                    self.buffered_responses.push(OrderedHeadersResponse {
+                        headers,
+                        request,
+                        retries,
+                        peer_id,
+                    })
+                }
+
+                Ok(())
+            }
+            // most likely a noop, because this error
+            // would've been handled by the fetcher internally
+            Err(err) => {
+                Err(HeadersResponseError { request, retries, peer_id: None, error: err.into() })
+            }
+        }
+    }
+
+    /// Handles the error of a bad response
+    ///
+    /// Returns the error if the request ran out if retries.
+    fn on_headers_error(&mut self, err: HeadersResponseError) -> Option<DownloadError> {
```

Comment:
we don't want to propagate error to the requester any more, would be better to handle it internally + move to infinite retries

Reply:
sgtm, can remove retry handling entirely?

Reply:
yup